tso-server/server/server.go

```diff
@@ -131,60 +204,240 @@ type session struct {
+	} else {
+		tso.zkConn, err = zkhelper.ConnectToZkWithTimeout(cfg.ZKAddr, zkTimeout)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+	}
+
 	tso.conns = map[net.Conn]struct{}{}
 
 	return tso, nil
 }
 
-// Run runs the timestamp oracle server.
-func (tso *TimestampOracle) Run() {
+type tsoTask struct {
+	tso *TimestampOracle
+
+	connCh chan net.Conn
+
+	interruptCh chan error
+	interrupted int64
+
+	stopCh chan struct{}
+}
+
+func (t *tsoTask) Run() error {
+	if t.Interrupted() {
+		return errors.New("task has been interrupted already")
+	}
+
+	// enter here means I am the leader tso.
+	tso := t.tso
+
+	log.Debugf("tso leader %s run task", tso)
+
+	if err := tso.loadTimestamp(); err != nil {
+		// should we interrupt the task here now?
+		return errors.Trace(err)
+	}
+
+	atomic.StoreInt64(&tso.isLeader, 1)
+
+	tsTicker := time.NewTicker(time.Duration(updateTimtestampStep) * time.Millisecond)
+
+	defer func() {
+		atomic.StoreInt64(&tso.isLeader, 0)
+
+		tsTicker.Stop()
+
+		// we should close all connections.
+		t.closeAllConns()
+		log.Debugf("tso leader %s task end", tso)
+	}()
+
+	for {
+		var conn net.Conn
+
+		select {
+		case err := <-t.interruptCh:
+			log.Debugf("tso leader %s is interrupted, err %v", tso, err)
+
+			// we will interrupt this task, and we can't run this task again.
+			atomic.StoreInt64(&t.interrupted, 1)
+			return errors.Trace(err)
+		case <-t.stopCh:
+			log.Debugf("tso leader %s is stopped", tso)
+
+			// we will stop this task, maybe run again later.
+			return errors.New("task is stopped")
+		case conn = <-t.connCh:
+			// handle connection below
+
+			s := &session{
+				r:    bufio.NewReaderSize(conn, sessionReadBufferSize),
+				w:    bufio.NewWriterSize(conn, sessionWriteBufferSize),
+				conn: conn,
+			}
+
+			tso.addConn(conn)
+			tso.wg.Add(1)
+			go tso.handleConnection(s)
+		case <-tsTicker.C:
+			if err := tso.updateTimestamp(); err != nil {
+				return errors.Trace(err)
+			}
+		}
+	}
+}
+
+func (t *tsoTask) Interrupted() bool {
+	return atomic.LoadInt64(&t.interrupted) == 1
+}
+
+func (t *tsoTask) Stop() {
+	select {
+	case t.stopCh <- struct{}{}:
+	default:
+		log.Warnf("task stop blocked")
+	}
+}
+
+func (t *tsoTask) closeAllConns() {
+	t.tso.closeAllConns()
+
+	// we may close the connection in connCh too
+	n := len(t.connCh)
+	for i := 0; i < n; i++ {
+		conn := <-t.connCh
+		conn.Close()
+	}
+}
+
+func (t *tsoTask) Listening() {
+	tso := t.tso
+
+	defer tso.wg.Done()
+
 	for {
 		conn, err := tso.listener.Accept()
 		if err != nil {
+			t.interruptCh <- errors.Trace(err)
 			return
 		}
 
-		s := &session{
-			r:    bufio.NewReaderSize(conn, sessionReadBufferSize),
-			w:    bufio.NewWriterSize(conn, sessionWriteBufferSize),
-			conn: conn,
+		if !tso.IsLeader() {
+			// here mean I am not the leader now, so close all accpeted connections.
+			conn.Close()
+			continue
 		}
 
-		tso.conns[conn] = struct{}{}
-		tso.wg.Add(1)
-		go tso.handleConnection(s)
+		t.connCh <- conn
```

Comment:
Seems use atomic for isLeader is not enough. `conn` may be sent to `cnnCh` after `t.closeAllConns()`.